### ISSUE TITLE
any/all are exposed as if they were real terms

### ISSUE THREAD

**Issue by marshall007** [rethinkdb/rethinkdb#3511]
If you run `r.expr(true).all()` in the data explorer instead of throwing, the query appears to be evaluated as  `r.expr(true).and()` and returns `true`. So the issue is two-fold:

1. I would expect to get `TypeError: undefined is not a function`
1. I think calling `and/or` with no arguments should throw something like `RqlDriverError: Expected at least 1 argument but found 0.`

Out of curiosity, I tried doing the same thing with a few other terms that are used internally but not exposed like `FUNC`, but it seems to only apply to `ANY/ALL`.

**Comment by danielmewes** [comment]
Can you explain a little bit more why this is a problem?
Currently `any` and `all` are in fact real terms. Do you think they should have different semantics?

I agree that `and`/`or` with only one argument (counting the one that's left of the term, in this case `r.expr(true)`) doesn't make much sense. I'm not sure if there's enough reason to disallow it though.

Maybe part of the confusion comes from the fact that we occasionally use the two names `any` and `or` / `all` and `and` interchangeably. I'm not entirely sure what the background of this is (@mlucy probably knows).
It seems that our API documentation only uses the `and`/`or` form nowadays and doesn't even mention `any`/`all`. I could imagine that we had `any`/`all` originally but switched over to `and`/`or` at some point.

**Comment by marshall007** [comment]
@danielmewes I didn't realize they were intended to be aliases to `and/or` since I couldn't find them in the docs. My expectation before consulting the docs was that `any` would be equivalent to `contains` and `all` would do something like [`Array.every()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/every) in JS.

**Comment by deontologician** [comment]
That was the reason they were de-emphasized. Because people expected them to act like any/all in python and other languages, and they don't.

@danielmewes I don't know if there is an issue for it already but maybe removing any/all before 2.0 is another good cleanup item? @mlucy might have thoughts on this as well

**Comment by danielmewes** [comment]
Thanks @marshall007 and @deontologician for your explanations.

I like the idea of getting rid of `any`/`all` for 2.0, unless there are any strong objections.

**Comment by danielmewes** [comment]
Marking as RQL_proposal. Note that the proposal here is to remove `any` and `all`, nothing else for now.

**Comment by mlucy** [comment]
`any` and `all` are considered "real" terms because they used to be the standard way to access `or` and `and`.  We did this because `or` and `and` are reserved in some languages, but `any` and `all` turned out to be confusing so we standardized on `or` and `and` instead.  I would also be in favor of remove `any` and `all` in 2.0; the only reason we haven't done it is to avoid breaking existing code.  (I would also possibly be on board with repurposing them to mean what they usually mean in Ruby.)

We let `and` and `or` take any number of arguments because it makes it easier to write generic code where you splice an array into a call to `r.and(...)`.
